### PR TITLE
NTDownloader UI revamp

### PR DESCRIPTION
**Pull request by MTandi**
## About The Pull Request

Revamp of the software downloader program for modular computers.

Changes:
- Programs are now sorted alphabetically with incompatible ones in the end of the list.
- Installed programs are now displayed in the list.
- Added program icons.
- Moved the error messages in place of the download button.
- Only the most important error message is displayed now. Priority: compatibility, access, free space. 
- Syndicate programs are now displayed in the same list, but have a warning message from NT (There are no warning messages on syndicate OS).
- Added program categories to improve navigation. The default option "All" contains items from all categories.
- Download progress bar moved in place of the Disk usage bar. Disk usage is updated only after the download is complete, so the information was inaccurate during download. And the download bar now always visible regardless of selected category.
- The old download progress bar (next to the corresponding program) is replaced with "Downloading" indicator with a spinner.

### New UI

https://user-images.githubusercontent.com/3625094/107362787-46f0a680-6aea-11eb-9077-7ca6ad9d34a0.mp4

## Why It's Good For The Game

- The old version was hard to navigate because programs were sorted by filename.
- Error messages used vertical space and were sometimes unnecessary (I don't care about disk space or access if the program is incompatible with my device anyway)
- The old version didn't show installed programs, so you had to exit the program to see what you have installed.
- The old version grouped programs by availability. So if you found a program in a list that required access, and you inserted your ID, the list sorting was updating and you had to look for that program once again.
- Categories improve navigation when you need only software from particular category to do your job.

## Changelog
:cl:
refactor: Software Downloader program UI revamp
/:cl:

### CODE REVIEW

**Comment by stylemistake** [comment]
> The old download progress bar (next to the corresponding program) is replaced with "Downloading" indicator with a spinner.

That's a downgrade :/

**Comment by MTandi** [comment]
> That's a downgrade :/

Why? You can download only one program anyway. And I didn't remove it, I moved the progress bar to the top.

**Comment by stylemistake** [comment]
That old progress bar looked intuitive, had a good use of space (download button was useless for the whole time it was downloadig), and would work well if someone actually added a download queue, then all of them would turn into progress bars, and fill up in queue order. Your solution is a bit of a downgrade in user experience, because it places download into free/used space area, which is counterintuitive.

File: tgui/packages/tgui/interfaces/NtosNetDownloader.js
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/NtosNetDownloader.js:71]: Replace Flex with Stack.

Wrap Tabs with `<Section fill fitted>` element.

**Review by stylemistake — CHANGES_REQUESTED** [review]
I will point out more things that could've been done better, I would not merge it in this state.

File: tgui/packages/tgui/interfaces/NtosNetDownloader.js
Comment (MTandi) [reply tgui/packages/tgui/interfaces/NtosNetDownloader.js:71]: Thank you, I was referring to the other UI and didn't know about Stack. I'll clean things up and do something with the Downloading button.

Comment (MTandi) [reply tgui/packages/tgui/interfaces/NtosNetDownloader.js]: I did not wrap tabs assuming that your other PR will be merged

**Comment by MTandi** [comment]
- Reverted the download progress bar change
- Added downloading indicator (spinner icon) to the top with a tooltip.
- Fix: Disk usage now shows real usage during download, and changed the label slightly.
- Disabled download buttons now have a tooltip

**Comment by stylemistake** [comment]
Wait for #56721 to be merged, then tabs will look ok, and no actions will need to be taken.

File: code/modules/modular_computers/file_system/programs/ntdownloader.dm
Comment (stylemistake) [inline code/modules/modular_computers/file_system/programs/ntdownloader.dm:197]: You already have category property here, there is no need to have nested lists by category, i.e. `list(category = list(...))`

**Review by stylemistake — CHANGES_REQUESTED** [review]
I want you to do a few specific changes to your code before I can merge it.

First, sorting every second in ui_data on the server is expensive. Remove sorting on the server.

Second, simplify the code that builds the program list. You do not need to form a list for each category. Just add programs to a single, flat list, which you will then filter in TGUI based on the category property. You do not need any extra processing to the list. This will make your ui_data code very simple.

Third, this is how it's done in tgui:

```js
const programs = flow([
  // This filters the list to only contain programs with category
  // equal to selectedCategory
  filter(program => program.category === selectedCategory),
  // This sorts all programs in the lists by `name`
  sortBy(program => program.name),
])(data.programs);
```

You will need these imports:

```js
import { flow } from 'common/fp';
import { filter, sortBy } from 'common/collections';
```

Send list of categories as a simple list in ui_data:
```dm
data["categories"] = list(
  "Crew",
  "Engineering",
  ...
)
```

Do not send "All", because you can hardcode this logic into your UI, and disable the filter if "All" is selected. "All" has no purpose in dm code.

<details>
  <summary>Example of how to conditionally enable a "filter" in "flow":</summary>

```js
const programs = flow([
  selectedCategory !== 'All' && (
    filter(program => program.category === selectedCategory)
  ),
  sortBy(program => program.name),
])(data.programs);
```

</details>

Finally, rebase your PR to get new changes in master (improved Sections and Tabs), and update the screenshot. Hopefully I'm not asking too much.

**Comment by MTandi** [comment]
Updated the video in the first comment. Sorry for the confusion with rebase.

File: tgui/packages/tgui/interfaces/NtosNetDownloader.js
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/NtosNetDownloader.js:39]:
```suggestion
    sortBy(
      program => -program.compatible,
      program => program.filedesc),
```

**Review by stylemistake — APPROVED** [review]

**Comment by Rohesie** [comment]
Does this need testing, @stylemistake ?
I see you approved it but didn't merge.

**Pull request merged**
